src/vma/util/dm_context.cpp

```diff
@@ -0,0 +1,262 @@
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "vlogger/vlogger.h"
+#include "vma/util/dm_context.h"
+#include "vma/proto/mem_buf_desc.h"
+#include "vma/dev/ib_ctx_handler.h"
+
+#if defined(HAVE_INFINIBAND_MLX5_HW_H)
+#if defined(HAVE_IBV_DM)
+
+#define DM_MEMORY_MASK_4  3
+#define DM_MEMORY_MASK_64 63
+#define DM_ALIGN_SIZE(size, mask) ((size + mask) & (~mask))
+
+#undef  MODULE_NAME
+#define MODULE_NAME 		"dmc"
+#undef  MODULE_HDR
+#define MODULE_HDR MODULE_NAME "%d:%s() "
+
+#define dmc_logerr	__log_info_err
+#define dmc_logwarn	__log_info_warn
+#define dmc_logdbg	__log_info_dbg
+#define dmc_logfunc	__log_info_func
+
+dm_context::dm_context() :
+	m_p_dm_mr(NULL),
+	m_p_mlx5_dm(NULL),
+	m_p_ring_stat(NULL),
+	m_allocation(0),
+	m_used(0),
+	m_head(0)
+{};
+
+/*
+ * Allocate dev_mem resources
+ */
+bool dm_context::dm_allocate_resources(ib_ctx_handler* ib_ctx, ring_stats_t* ring_stats)
+{
+	size_t allocation_size = DM_ALIGN_SIZE(safe_mce_sys().ring_dev_mem_tx, DM_MEMORY_MASK_64);
+	struct ibv_exp_alloc_dm_attr dm_attr = {allocation_size, 0};
+	struct ibv_mr_attr mr_attr;
+	struct ibv_dm *ibv_dm;
+	m_p_ring_stat = ring_stats;
+
+	if (!allocation_size) {
+		// Device memory usage was disabled by the user
+		return false;
+	}
+
+	if (!ib_ctx->get_on_device_memory_size()) {
+		// Device memory usage is not supported
+		return false;
+	}
+
+	// Initialize attributes
+	bzero(&mr_attr, sizeof(mr_attr));
+	mr_attr.type = IBV_DEV_MEM;
+	mr_attr.length = allocation_size;
+	mr_attr.access = IBV_ACCESS_LOCAL_WRITE;
+	mr_attr.mem_type.dev_mem.dm = NULL;
+	mr_attr.mem_type.dev_mem.offset = 0;
+	mr_attr.comp_mask = 0;
+
+	// Allocate device memory buffer
+	ibv_dm = ibv_exp_alloc_dm(ib_ctx->get_ibv_context(), &dm_attr);
+	if (!ibv_dm) {
+		dmc_logerr("ibv_exp_alloc_dm() error - device memory allocation failed, %d %m", errno);
+		return false;
+	}
+
+	// Register device memory MR
+	mr_attr.mem_type.dev_mem.dm = ibv_dm;
+	m_p_dm_mr = ibv_reg_mr_ex(ib_ctx->get_ibv_pd(), &mr_attr);
+	if (!m_p_dm_mr) {
+		ibv_exp_free_dm(ibv_dm);
+		dmc_logerr("ibv_exp_free_dm error - dm_mr registration failed, %d %m", errno);
+		return false;
+	}
+
+	m_allocation = allocation_size;
+	m_p_mlx5_dm = reinterpret_cast<struct vma_mlx5_dm *> (ibv_dm);
+	m_p_ring_stat->n_tx_dev_mem_allocated = m_allocation;
+
+	dmc_logwarn("Device memory allocation completed successfully! device[%s] bytes[%zu] dm_mr handle[%d] dm_mr lkey[%d]",
```

Comment:
why warning?

Reply:
for debug reasons. of course it will not be merged as is now.